torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -367,8 +368,32 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
         );
       }
     }
-    // Flatten the tensors (from all ranks) into a single big tensor.
-    flattened[i] = newLikeFlat(tensor_lists, i);
+
+    //  Check if the tensors are already flattened to do inplace operation
+    bool inplace = true;
+    for (auto j = size_t{}; j < tensor_lists[i].size(); ++j) {
+        auto t = tensor_lists[i][j];
+        if (!t.storage().is_alias_of(other[i].storage()) ||
```

Comment:
I might be confused about this. Why outputs should share storage with inputs? Don't we just want all outputs to share storage or this is a special case of allocating storage where outputs are combined with inputs (I didn't know this is allowed by NCCL engine!). 
If that's the case, this saves some memory for separate input allocation in addition to output allocation!

However, for this specific case, maybe it's just enough to check if all outputs are properly aligned? 
Do you know what would happen if outputs are aligned and share storage with inputs but inputs aren't properly placed within outputs, would NCCL crash or produce garbage outputs?

Reply:
NCCL library allows inplace operations, it's explicitly mentioned in the documention. The inputs have to be properly aligned. I don't know what happens if they are not, other than it will not work

![image](https://user-images.githubusercontent.com/41591019/75594234-b55d6580-5a4d-11ea-91c8-207e00f78b0b.png)

Reply:
That makes sense. In-place will definitely save memory in addition to copying data. 

I was wondering if we can save copying in case we've allocated output tensors separately from inputs, but outputs are still contiguous (that is, already flat). Something like:

input = torch.tensor(size)
....
outputs = torch.tensor(size*world_size*num_devices)
pg.allgather(output, input)

Do you think this usecase is worth covering as well?

(Alternatively we can just ask users to copy 'input' at the appropriate location inside 'output' and change its storage accordingly.)

What do you think?

Reply:
Both cases should work. For all-gather, we can avoid copying to flatten and un-flatten if the outputs are views into an already flattened tensor. In addition, if the input is also a view into the same flattened tensor and input is properly aligned, we can do an in-place operation. The code in my PR already does this, but I don't verify that the input is correctly aligned. I will add code to verify proper alignment.